compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/AbstractCompareAndSwapNode.java

```diff
@@ -50,6 +50,7 @@
     @Input ValueNode expectedValue;
     @Input ValueNode newValue;
     @OptionalInput(State) FrameState stateAfter;
+    protected final boolean isJavaVolatile;
```

Comment:
I think I understand the practical meaning of this flag but I'm not sure the naming is clear enough.  So basically in ARM there are 2 ways of doing volatiles and this flag is meant to distinguish between these 2 cases?  In almost all cases Java volatile should be considered the normal one I think and maybe this flag should be something like isNonStandardVolatile?  Actually what are there cases where we use a CompareAndSwapNode against a location which is acquired using the other sequence?

Reply:
There are indeed two ways of generating compare and swap code. The choice of generation strategy depends on two factors: what location is being updated and what value has been set for a AArch64 flag (UseBarriersForVolatile).

If UseBarriersForVolatile is true then all compare and swap operations use ldxr and stxr combined with dmb instructions. 

If UseBarriersForVolatile is false then the nature of the memory location comes into play.

When the location is a Java field then ldaxr and stlxr need to be be used. This is consistent with the semantics employed for generating Java volatile reads and implemented in PR 1772 where they translate to ldar and stlr.

However, since PR has not yet been merged1772 XiahongGong has added an extra dmb at the end of ldax/stlxr sequence to deal with the disparity between the translation this patch implements and the current translation for volatiles a combination of ldr/str with dmb instructions.

When the location is not a Java field then the code must use ldxr, stxr combined with dmb instructions. This is consistent with normal expectations regarding memory synchronization  non-Java data.

Reply:
Ok, but why are the other locations that are using isJavaVolatile == false correct?  Are all of those clearly non-Java locations?  For instance in the context of SVM I think everything is a Java volatile.

Reply:
As I understand it (I hope XiaohongHong will correct me if I misrepresent things) XiaohongHong has only wired up the unsafe plugins for compare and exchange/swap to pass true. Those are cases that are known to correspond to a Java application program explicitly posting a read or write to Java data. The relaxing of the semanticsni these cases is compatible with the optimization implemented by PR 1772 when UseBarriersForVolatile is false and improves performance. Or rather it wil be when 1772 is merged and we can omit the tarailing barriers that are stil in place for this patch.

In other cases where code internal to the JIT or Substrate creates an UnsafeCompareAndSwapNode or UnsafeCompareAndSwapNode via the original constructor the value defaults to false which means the generator uses a more heavyweight barrier-based translation. This alternative is always consistent (on all known AArch64 arches) whether or not PR 1772 is in place. It is just less performant.

The reason the other cases have not been changed to use ldaxr/stlxr  is because it was not clear whether these internal uses were expecting other accesses to the target locations to be simple Java volatile field reads and writes (which would be correctly handled by the JIT once PR 1772 was merged) or to require the full semantics that is normally expected for compare and swap. The idea was to leave them as is and then convert the relevant uses one by one as needed. If you can confirm that all cases where the existing code explicitly constructs an UnsafeCompareAndSwapNode or UnsafeCompareAndSwapNode are compatible with PR 1772 then we don't need to worry about the isJavaVolatile == false case and can always use this optimization.

Reply:
Yes that's what I'm getting at.  So for instance does the CompareAndSwapNode that we emit in the MonitorSnippets using Pointer.compareAndSwapWord need to use the conservative pattern because HotSpot uses a different pattern or can it use the more optimal pattern?  In the context of this PR it might be helpful to eliminate the constructor that passes the default value to make it more clear where we have to think about the setting of this flag.
In the context of SVM it's really only talking to itself I think so I would guess that always using the optimal pattern would be possible though it could necessary to have an escape hatch.
Apart from the Unsafe plugin, selecting the kind of volatile based on the LocationIdentity might work fairly well.  In the MonitorSnippets case for instance the MARK_WORD_LOCATION could be used to select the right kind for Hotspot if it needs something different but still do the normal thing for SVM.  Anyway, I'd like to get more clarity on which cases we actually need.

Reply:
We mainly concerns the usages in SVM, like in `JNIJavaVM.addJavaVM` and others in isolated, and its future usages in the internal VM of native image like GC. I'm not very much sure about where we should add the conservative barriers. But just in case, I think we should add a choice for the developer.
For the `MonitorSnippets `in HotSpot, interpreter and C1 always uses the conservative model while C2 uses the same one with Graal now. 
For the `isJavaVolatile` flag, do you mean we should eliminate the constructor that will set the default value to it, or just delete it because we could also decide the memory model based on the `LocationIdentity`?

Reply:
>  Anyway, I'd like to get more clarity on which cases we actually need

Well, yeah, me too ;-) Only I'm not really in a position to determine that. That's why the current change was chosen: to be conservative and minimally invasive while still allowing further optimizations to be applied gradually as opportunities were  identified.

You are right that compilation for Substrate can probably always use the optimized model and that includes, say, the MonitorSnippets case. However, that code is currently shared with the Hotspot generation case. I'm not really clear how to separate those cases and apply the right transformation and I don't believe XiahongGong is. Advice would be very welcome.

Reply:
> At AArch64LIRGenerator:178 you have misspelled method name 'emitComapreAndSwap'. Calls also need correcting.

Fixed in the updated version! Thanks!

Reply:
Hi @tkrodriguez , I have renamed the `isJavaVolatile` flag to `useBarriers` and eliminated the constructor that sets the default value to it. Besides, I did the modifications compared to the last version:
1. Add a `conservative` field for `LocationIdentity` to decide whether to use the full semantic barriers for VM internal atomic operations. I only set the flag for the CAS used `LocationIdentity `in `MonitorSnippets` currently. People can set the flag for other atomic operations if needed in future.
2. Remove the reference of `UseBarriersForVolatile`, always use the optimized `ldaxr/stlxr `pattern for java Unsafe atomics.

Could you please take a look whether this version is more reasonable? Thanks!

Reply:
The reason HotSpot uses the "conservative" memory barriers is that it's very old code, and no-one could guarantee that "merely" sequentially-consistent operations would be adequate. However, I believe that SC is really adequate for HotSpot and we quietly used such operations in the AArch64 port for years until someone noticed. Graal shouldn't need anything more than SC, and if it does anywhere, that's a bug in Graal. IMO.

Reply:
Hi @theRealAph, thanks for looking at this PR. So do you mean "ldaxr/stlxr" is SC enough for vm internal atomics like monitor operations in Graal?  Actually we didn't met any issue for it.

Reply:
Yes, that's what I mean. However, it is possible that there is (broken) code elsewhere in Graal which assumes monitors, CAS, etc. have full-barrier semantics. I have no idea how to find such code.

Reply:
Yes, I'm also worried about the possible scenario and actually have no idea to make a decision. However, the main purpose of this PR is to add the full-barrier alternative implementation for CAS, so that people can have a chance to use it if needed in future.  If there is not any known issue for the current pattern `(ldaxr/stlxr)` used in hotspot monitor (since we have used it for long times in C2), I think we can keep using it currently in Graal. But it's still worth to add the full-barrier alternative implementation in Graal.